lib/dynflow/director.rb

```diff
@@ -42,33 +64,70 @@ def initialize(execution_plan_id, step, queue)
       def execute
         @step.execute(nil)
       end
+
+      def to_hash
+        super.merge(step: step.to_hash)
+      end
+
+      def self.new_from_hash(hash, *_args)
+        self.new(hash[:execution_plan_id],
+                 Serializable.from_hash(hash[:step], hash[:execution_plan_id], Dynflow.process_world),
+                 hash[:queue])
+      end
     end
 
     class EventWorkItem < StepWorkItem
-      attr_reader :event
+      attr_reader :event, :request_id
 
-      def initialize(execution_plan_id, step, event, queue)
+      def initialize(request_id, execution_plan_id, step, event, queue)
         super(execution_plan_id, step, queue)
         @event = event
+        @request_id = request_id
       end
 
       def execute
-        @step.execute(@event.event)
+        @step.execute(@event)
+      end
+
+      def to_hash
+        super.merge(request_id: @request_id, event: Dynflow.serializer.dump(@event))
+      end
+
+      def self.new_from_hash(hash, *_args)
+        self.new(hash[:request_id],
+                 hash[:execution_plan_id],
+                 Serializable.from_hash(hash[:step], hash[:execution_plan_id], Dynflow.process_world),
+                 Dynflow.serializer.load(hash[:event]),
+                 hash[:queue])
       end
     end
 
     class FinalizeWorkItem < WorkItem
-      def initialize(execution_plan_id, sequential_manager, queue)
+      attr_reader :finalize_steps_data
+
+      # @param finalize_steps_data - used to pass the result steps from the worker back to orchestrator
+      def initialize(execution_plan_id, queue, finalize_steps_data = nil)
         super(execution_plan_id, queue)
-        @sequential_manager = sequential_manager
+        @finalize_steps_data = finalize_steps_data
       end
 
       def execute
-        @sequential_manager.finalize
+        execution_plan = world.persistence.load_execution_plan(execution_plan_id)
+        manager = Director::SequentialManager.new(world, execution_plan)
+        manager.finalize
+        @finalize_steps_data = manager.finalize_steps.map(&:to_hash)
+      end
+
+      def to_hash
+        super.merge(finalize_steps_data: @finalize_steps_data)
+      end
+
+      def self.new_from_hash_(hash, *_args)
```

Comment:
To fix execution plans with finalize steps ending in stopped-pending do `sed 's/_(/(/'` on this line

Reply:
Sorry, my bad: I was just testing that the code is not working without the method (and you've noticed it wasn't), and forgot to put it back. Updated